### PR TITLE
[files/build/versions]: support reproduceable build for git

### PR DESCRIPTION
**Pull request by liushilongbuaa**
**- Why I did it**
Some component's make process need thirdparty source file which is downloaded by "git clone", not using a specific commit. When source file is changed, Build process may fail. 
**- How I did it**
1. Add a version file, files/build/versions/versions-git, to record thirdparty source code version(commit ID).
2. Replace git binary by a bash script. In the script, when 'git clone' is done, 'git reset' will reset HEAD to the commit recorded in the version file.
3. When update version file, you need to set parameter 'SONIC_VERSION_CONTROL_COMPONENTS ' and build. At last replace 'files/build/versions/versions-git' by 'files/build/versions/versions-git-latest'
**- How to verify it**
'files/build/versions/versions-git.log' records some infomation.
**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [ ] 201911
- [ ] 202006

**- Description for the changelog**

Control thirdparty source code version used by 'git clone'

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: Makefile.work [hunk 1]
```diff
@@ -38,6 +38,13 @@
 #  * SONIC_DPKG_CACHE_METHOD: Specifying method of obtaining the Debian packages from cache: none or cache
 #  * SONIC_DPKG_CACHE_SOURCE: Debian package cache location when cache enabled for debian packages
 #  * BUILD_LOG_TIMESTAMP: Set timestamp in the build log (simple/none)
+#  * SONIC_VERSION_CONTROL_COMPONENTS : In the process of SONiC build, there are some source code,
```
Comment (jleveque) [inline Makefile.work:41]: Suggest removing `SONIC_` prefix from user-configurable variable for brevity.

Comment (liushilongbuaa) [reply Makefile.work:41]: thanks~

**Comment by liushilongbuaa** [comment]
restart all please

**Comment by liushilongbuaa** [comment]
retest broadcom please

File: files/build/scripts/git [hunk 2]
```diff
@@ -0,0 +1,86 @@
+#!/bin/bash
+
```
Comment (qiluo-msft) [inline files/build/scripts/git:2]: Please add file header comment to explain the purpose and usage.

File: Makefile.work [hunk 3]
```diff
@@ -38,6 +38,13 @@
...
+#  *                               which are downloaded by third utils. ex: git,wget,pip,docker...
+#  *                    By default build will use version file in 'files/build/versions/' to control third utils version.
+#  *                    if we specify one component, build will use latest version and record it in a new file
+#  *                    ex: git's version file is files/build/versions/versions-git, if ${var} is "pip,git"
+#  *                        git clone will use the latest commit, and copy the commit ID to files/build/versions/versions-git-latest.
+#  *                        When build succeeded, you can replace versions-git by versions-git-latest, and commit the changes.
 #
```
Comment (qiluo-msft) [inline Makefile.work:48]: What about the `Default` and `Values` fields?

File: files/build/versions/versions-git [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+https://github.com/Mellanox/libpsample.git==e48fad2402500a2aeb73961729aad06f793ff46e
+https://github.com/sflow/host-sflow==bdf424b3f1c542a665dc8e60233802028eb7addb
+https://github.com/sflow/sflowtool==71bf6ce2e3dac3a1621527de5c8885283fe406cf
+https://github.com/Supervisor/supervisor.git==d0b583e25f76bcbb079398cfba39bcbfb1971cdc
+https://github.com/thom311/libnl==2abeec851e4c748effac341f8db0b42389b37790
+https://salsa.debian.org/debian/libteam.git==f8808df228b00873926b5e7b998ad8b61368d4c5
+https://salsa.debian.org/debian/python-click==da0b4b204db67f708771614a7d133581f6d02065
+https://salsa.debian.org/dhcp-team/isc-dhcp.git==d0d58776805d720029619ea06df5c57706eecd5a
+https://salsa.debian.org/kernel-team/initramfs-tools.git==40e544e13611c1b2690eb99a8096fc16c1b9c74e
+https://salsa.debian.org/sk-guest/monit.git==8b897931d4b86a1b4af5052a796e60b67344dfdb
```
Comment (qiluo-msft) [inline files/build/versions/versions-git:15]: >https [](start = 0, length = 5)

Pending comment on design doc: [https://github.com/Azure/SONiC/pull/684#discussion_r530079788](https://github.com/Azure/SONiC/pull/684#discussion_r530079788)

File: files/build/scripts/git (hunk 2 above)
Comment (liushilongbuaa) [reply files/build/scripts/git:2]: add header comment

File: Makefile.work (hunk 1 above)
Comment (liushilongbuaa) [reply Makefile.work:41]: keep prefix for consistent with existing parameter. this parameter mainly used in configure file.

**Comment by liushilongbuaa** [comment]
retest vs please

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Comment by liushilongbuaa** [comment]
retest mellanox please

File: files/build/versions/versions-git (hunk 4 above)
Comment (liushilongbuaa) [reply files/build/versions/versions-git]: replaced by ','

File: Makefile.work (hunk 3 above)
Comment (liushilongbuaa) [reply Makefile.work:50]: changed

**Comment by liushilongbuaa** [comment]
retest vs please

**Comment by liushilongbuaa** [comment]
retest baseimage please

**Comment by liushilongbuaa** [comment]
retest vs please

**Comment by liushilongbuaa** [comment]
retest broadcom please

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Comment by liushilongbuaa** [comment]
retest vsimage please

File: src/sonic-build-hooks/hooks/git [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        # record version file
+        echo "$URL==$commit_latest" >> $new_version_file
+        sort $new_version_file -o $new_version_file -u &>> /dev/null
+    fi
+    popd &> /dev/null
+
+    exit $result
+}
+
+main $@
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:93]: Change to "$@" ? Do you test the command parameter contains space characters?

File: src/sonic-build-hooks/hooks/git [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    whitespace="[[:space:]]"
+    final_str=()
+
+    # parse input parameters
+    for i in "$@"
+    do
+        if [[ $MODECOMMIT == 1 ]];then
+            if [[ $i =~ $whitespace ]];then
+                i=`echo $i | sed 's/ /_/g'`
+            fi
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:21]: Why replace the space to _?

File: Makefile.work [hunk 7]
```diff
@@ -104,8 +104,8 @@ $(shell CONFIGURED_ARCH=$(CONFIGURED_ARCH) j2 $(SLAVE_DIR)/Dockerfile.j2 > $(SLA
 $(shell BUILD_SLAVE=y scripts/prepare_docker_buildinfo.sh $(SLAVE_BASE_IMAGE) $(SLAVE_DIR)/Dockerfile $(CONFIGURED_ARCH) "" $(BLDENV))
 
 # Add the versions in the tag, if the version change, need to rebuild the slave
-SLAVE_BASE_TAG = $(shell cat $(SLAVE_DIR)/Dockerfile $(SLAVE_DIR)/buildinfo/versions/versions-* | sha1sum | awk '{print substr($$1,0,11);}')
-SLAVE_TAG = $(shell cat $(SLAVE_DIR)/Dockerfile.user $(SLAVE_DIR)/Dockerfile $(SLAVE_DIR)/buildinfo/versions/versions-* | sha1sum | awk '{print substr($$1,0,11);}')
+SLAVE_BASE_TAG = $(shell cat $(SLAVE_DIR)/Dockerfile $(SLAVE_DIR)/buildinfo/versions/versions-* src/sonic-build-hooks/hooks/* | sha1sum | awk '{print substr($$1,0,11);}')
+SLAVE_TAG = $(shell cat $(SLAVE_DIR)/Dockerfile.user $(SLAVE_DIR)/Dockerfile $(SLAVE_DIR)/buildinfo/versions/versions-* src/sonic-build-hooks/hooks/* | sha1sum | awk '{print substr($$1,0,11);}')
```
Comment (xumia) [inline Makefile.work:108]: The change will be not necessary, there is a fix for it, see: https://github.com/Azure/sonic-buildimage/pull/6348

File: src/sonic-build-hooks/hooks/git (hunk 5 above)
Comment (liushilongbuaa) [reply src/sonic-build-hooks/hooks/git:93]: changed to "$@", and test in bash

File: src/sonic-build-hooks/hooks/git (hunk 6 above)
Comment (liushilongbuaa) [reply src/sonic-build-hooks/hooks/git:21]: fixed with quotation mark.

File: Makefile.work (hunk 7 above)
Comment (liushilongbuaa) [reply Makefile.work]: removed

**Comment by liushilongbuaa** [comment]
retest broadcom please

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Comment by liushilongbuaa** [comment]
retest vsimage please

File: src/sonic-build-hooks/hooks/git [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    [ -f $version_file ] && commit=`grep $URL $version_file | awk -F, '{print$2}'`
+
+    # control version or record version file
+    if [[ $ENABLE_VERSION_CONTROL_GIT == "y" ]];then
+        # control version
+        [ ! -z $commit ] && $REAL_COMMAND reset --hard $commit &>> /dev/null
+    else
+        # record version file
+        echo "$URL==$commit_latest" >> $new_version_file
+        sort $new_version_file -o $new_version_file -u &>> /dev/null
```
Comment (qiluo-msft) [inline src/sonic-build-hooks/hooks/git:69]: >&>> [](start = 55, length = 3)

Why sometime using `&>>`, but other using `&>`? Possible to unify?

**Comment by liushilongbuaa** [comment]
retest vsimage please

**Review by xumia — APPROVED** [review]

File: src/sonic-build-hooks/hooks/git [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        esac
+    done
+
+    # if not specific clone path, get default clone path
+    [ -z $clone_PATH ] && clone_PATH=`echo $URL | rev | awk -F/ '{print$1}' | rev | awk -F. '{print$1}'`
+}
+
+main(){
+    parse_config "$@"
+    get_clone_path "$@"
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:47]: Can be moved to line 57

**Review by qiluo-msft — APPROVED** [review]
LGTM. Please also check with other reviewers.

**Pull request merged**